### PR TITLE
Mavens getting listed in idle,idle7d+

### PR DESCRIPTION
**Pull request by fakhruddinkw**
Date: 06-09-2024 

Developer Name:  Fakhruddin KW

## Issue Ticket Number
#2024

## Description
In this PR, we have collected all the users with role maven and preventing them from getting group-idle,group-idle-7d+,onboarding-31d

onboarding-31d was already done, this PR takes care of group-idle,group-idle-7d+

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

https://github.com/user-attachments/assets/c117da92-536f-4281-b448-b4aa5b3baa3b

## Screenshots

</details>

## Test Coverage
group-idle

group-idle-7d

**BEFORE**:
discord-coverage

**AFTER**:
discord-coverage-1

### CODE REVIEW

File: models/discordactions.js [hunk 1]
```diff
@@ -349,7 +355,7 @@ const updateIdleUsersOnDiscord = async () => {
             if (userData.exists) {
               if (isUserArchived) {
```
Comment (iamitprakash) [inline models/discordactions.js:356]: please re-factor nested if

File: models/discordactions.js [hunk 2]
```diff
@@ -549,6 +555,7 @@ const updateIdle7dUsersOnDiscord = async () => {
   let allUsersHavingGroupIdle7d = [];
   let groupIdle7dRole;
   let groupIdle7dRoleId;
+  const allMavens = [];
```
Comment (iamitprakash) [inline models/discordactions.js:558]: can't we use from line 327 or make it global

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: models/discordactions.js (hunk 2 above)
Comment (fakhruddinkw) [reply models/discordactions.js:558]: the thing is, these APIs are called in parallel, so i am not sure how i would do that? can you please help?

**Comment by sahsisunny** [comment]
@fakhruddinkw test is not required?

**Comment by fakhruddinkw** [comment]
> @fakhruddinkw test is not required?

they are, I am working on them. I have made the PR open for review if changes are required

**Comment by tejaskh3** [comment]
Please add test coverage, also the changes you have done doesn't have any test cases written for them.

**Review by tejaskh3 — CHANGES_REQUESTED** [review]
Please add test coverage, also the changes you have done doesn't have any test cases written for them.

**Comment by fakhruddinkw** [comment]
> Please add test coverage, also the changes you have done doesn't have any test cases written for them.

on it

File: test/integration/tasks.test.js [hunk 3]
```diff
@@ -1516,7 +1516,7 @@ describe("Tasks", function () {
       expect(response.body).to.be.deep.equal({
         message: "Discord details of users with status missed updates fetched successfully",
         data: {
-          usersToAddRole: [activeUserWithProgressUpdates.discordId],
+          usersToAddRole: [],
```
Comment (vinit717) [inline test/integration/tasks.test.js:1519]: Why it is empty?
and add test case related to maven part

File: test/integration/discordactions.test.js [hunk 4]
```diff
@@ -615,9 +616,30 @@ describe("Discord actions", function () {
...
+        ...userData[2],
+        discordId: "123456",
+        discordJoinedAt: "2023-07-31T16:57:53.894000+00:00",
+        roles: { archived: false, in_discord: true },
+      };
+
+      getDiscordMembers[1] = {
+        ...getDiscordMembers[1],
+        roles: ["1212121212"],
+      };
```
Comment (vinit717) [inline test/integration/discordactions.test.js:641]: Can you add this is fixture if other test is not failing

File: models/discordactions.js
Comment (vinit717) [inline models/discordactions.js:1]: Please add model test also

**Review by vinit717 — CHANGES_REQUESTED** [review]

File: test/integration/tasks.test.js (hunk 3 above)
Comment (fakhruddinkw) [reply test/integration/tasks.test.js]: I have fixed the test and have reverted back to old code

File: test/integration/discordactions.test.js (hunk 4 above)
Comment (fakhruddinkw) [reply test/integration/discordactions.test.js]: I had to do this way as there were other tests failing

File: models/discordactions.js
Comment (fakhruddinkw) [reply models/discordactions.js:1]: Hi Vinit, valid point. but can i make another issue for this as the model tests were not written prior to my task?

File: models/discordactions.js [hunk 5]
```diff
@@ -324,6 +325,7 @@ const updateIdleUsersOnDiscord = async () => {
   let allUsersHavingGroupIdle = [];
   let groupIdleRole;
   let groupIdleRoleId;
+  allMavens = [];
```
Comment (iamitprakash) [inline models/discordactions.js:328]: do we really need this line ..?

**Review by iamitprakash — CHANGES_REQUESTED** [review]

**Comment by iamitprakash** [comment]
Hello @fakhruddinkw  please get peer review first, before coming to me

File: test/fixtures/discordResponse/discord-response.js [hunk 6]
```diff
@@ -68,6 +68,52 @@ const getDiscordMembers = [
...
+  {
+    avatar: "",
+    communication_disabled_until: "",
+    flags: 0,
+    is_pending: false,
+    joined_at: "some_intensional_date", // ISO8601 timestamp
+    nick: "goliath",
+    pending: false,
+    premium_since: null,
+    roles: ["giant", "hero"],
```
Comment (iamitprakash) [inline test/fixtures/discordResponse/discord-response.js:80]: move required role to this test data collection

File: test/integration/discordactions.test.js [hunk 7]
```diff
@@ -956,4 +978,80 @@ describe("Discord actions", function () {
...
+        })
+      );
+    });
+
+    afterEach(async function () {
+      sinon.restore();
+      await cleanDb();
+    });
+
+    it("should update Idle Users successfully and return a 201 status code", function (done) {
```
Comment (iamitprakash) [inline test/integration/discordactions.test.js:1039]: this 201 code looks confusing to me

File: test/integration/discordactions.test.js [hunk 8]
```diff
@@ -956,4 +978,80 @@ describe("Discord actions", function () {
...
+      chai
+        .request(app)
+        .put(`/discord-actions/group-idle`)
+        .set("Cookie", `${cookieName}=${superUserAuthToken}`)
+        .end((err, res) => {
+          if (err) {
+            return done(err);
+          }
+          expect(res).to.have.status(201);
```
Comment (iamitprakash) [inline test/integration/discordactions.test.js:1048]: same here

File: test/unit/models/discordactions.test.js [hunk 9]
```diff
@@ -801,4 +811,184 @@ describe("discordactions", function () {
...
+        })
+      );
+    });
+
+    afterEach(async function () {
+      sinon.restore();
+      await cleanDb();
+    });
+
+    it("should return  totalIdleUsers as 3,totalRolesApplied as 3, totalRoleToBeAdded as 3 as no one is maven", async function () {
```
Comment (iamitprakash) [inline test/unit/models/discordactions.test.js:986]: test description looks confusing to me, can we update please

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: test/integration/discordactions.test.js (hunk 7 above)
Comment (fakhruddinkw) [reply test/integration/discordactions.test.js:1039]: This was an old test which i have edited, i will look into it

File: test/integration/discordactions.test.js (hunk 8 above)
Comment (fakhruddinkw) [reply test/integration/discordactions.test.js:1048]: same as above

File: test/unit/models/discordactions.test.js (hunk 9 above)
Comment (fakhruddinkw) [reply test/unit/models/discordactions.test.js:986]: on it

File: models/discordactions.js (hunk 5 above)
Comment (fakhruddinkw) [reply models/discordactions.js:328]: no, will remove it.

File: test/fixtures/discordResponse/discord-response.js (hunk 6 above)
Comment (fakhruddinkw) [reply test/fixtures/discordResponse/discord-response.js:80]: I had to add roles separately in order to avoid messing with other existing tests

**Comment by fakhruddinkw** [comment]
> Hello @fakhruddinkw please get peer review first, before coming to me

Sorry my mistake, i will keep this in mind for future.

**Review by vinit717 — APPROVED** [review]

**Review by sahsisunny — APPROVED** [review]

**Pull request merged**